### ISSUE TITLE
Drone fuel economy is wrong 

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#711]
**Describe the bug**
In a Delivery mission the fuel required for a Drone is always too short of the required for the actual Mission. The distance calculates are correct so the problem is in the fuel economy calculations.

**To Reproduce (Please provide steps)**
Connect to a Settlemnt in the console and user ```create mission``` to create a mission where the distance is more than 200KM.
The Drone will run out of fuel @ ~80% distance.

**Additional context**
A refined follow on Issue from #664

**Comment by mokun** [comment]
Do you think there's a fuel estimation error ? 

Other than that, I can think of that there's a weight limit on how fuel a drone can carry. 

On top of that, it needs to carry the fuel oxidizer, namely, the oxygen onboard as well.

On earth, the oxygen is free available in the atmosphere, thus it saves the weight of carrying it. 

Not so with the atmosphere of Mars where oxygen is minimal.

Also, the output product as a result of the burning of the fuel is water (H2O). 

We could remove these generated H2O to lessen the weight during the flight. I haven't done that.

**Comment by bevans2000** [comment]
I debugged the start of a Delivery mission and for the needed fuel calculations the distance is correct.  So the economy used before flight is less than that used when moving, hence it burns fuel quicker than expected.

**Comment by mokun** [comment]
> So the economy used before flight is less than that used when moving

ok. Then we need to come up with a better estimated fuel economy figure.

**Comment by bevans2000** [comment]
It is probably due to the weight so before flight we should assume maximum weight?

**Comment by mokun** [comment]
> It is probably due to the weight so before flight we should assume maximum weight?

Yes. it was due to the water accumulated onboard from fuel cell reactions. It wasn't accounted for in the past.

We could modify the calculation of the initial estimate to include water. 

Also, during the flight, it can constantly check if there will be enough fuel to get to the destination, if not, see at what point it has to start shedding some water to lighten the load intelligently.

If it can arrive with all the water onboard, it can be transferred to the destination settlement for receiving a credit since water should be still considered a precious commodity.

**Comment by mokun** [comment]
Ever since adding that new conservative fuel economy figure, it requires more fuel and oxygen for each mission and the max mission range has since decreased. 

Does it achieve the effect we want ? 

May be we should decouple the fuel use and the max mission range ?!

**Comment by Urwumpe** [comment]
I think the effect is as expected, but we should fix the fuel cell reactions. They should be at constant mass, unless we dump water overboard. The water produced by the reaction has the same mass as the fuel consumed. 

If you want to extend range there with that technology, I only see the option to dump the water overboard, but as water is really precious, I wouldn't make it the default behaviour. 

For longer ranges than 200 km, we could attempt rocket mail (semiballistic trajectory). Mars thin atmosphere would be at least useful for something then.

**Comment by mokun** [comment]
> They should be at constant mass, unless we dump water overboard. The water produced by the reaction has the same mass as the fuel consumed.

Okay. But because it's going through chemical reactions, the input mass and the output mass will be related as follows : 
 
![1286267_1332356_ans_90e473b673df40fd8920a8faa1bf3bec](https://user-images.githubusercontent.com/1168584/232256488-115d4fcf-9f60-4ad7-b7a7-8e4b9bc5496f.jpg)

Given : 
1. input mass of Methane = 16 g
2. input mass of 2 moles of oxygen = 64 g
3. output mass of carbon dioxide = 44 g
4. output mass of 2 moles of water = 36 g

The thing is that we assume CO2 generated (44** g) is a waste gas and is being ejected.

But we could keep the generated water (36 g) onboard but to carry it during the flight would result in diminishing return since it will cost more fuel.

**Comment by mokun** [comment]
@Urwumpe 

Btw, since Mars has very little oxygen in the atmosphere, we don't have the luxury of the ambient oxygen in the air. 

So we'll also need to supply oxygen in order for the fuel cell reaction to take place.

Of course, this is also true on Mars if we want to resort to *burning* methane (instead of using fuel cell) to extract energy, we actually need to supply twice as much oxygen for the combustion to take place.

So realistically, it does seem to me that nuclear reactor is the best way to go to jumpstart an initial colony. 

Do you think so ?

**Comment by Urwumpe** [comment]
> @Urwumpe

> 
> So realistically, it does seem to me that nuclear reactor is the best way to go to jumpstart an initial colony.
> 
> Do you think so ?

I would differ, since there are many different aspects there.

In spaceflight applications, nuclear reactors have lost their performance advantage to solar up to Jupiter. And costs is also higher. But nuclear power could allow a higher sustained power, which in our case might be initially really an advantage, until a larger power supply is available.

Another disadvantage though: The reactor would need to use weapons-grade uranium to last long enough. Also we would have no water, so the reactor power would be limited by dry cooling methods. 

IF we could fit a relatively large reactor (RPV size, fuel inventory, see the PWR2 of the Astute class submarines, which is 30% larger than comparable reactors) with moderate power output AND the cooling system on a single lander, it could work. But it would likely be a dedicated mission. 

IF not, PV would be the way to go. Power would be more limited initially, but all needed equipment is small and modular enough to share their lander with other necessary gear. And later it would be one power source anyway.

But that is no solution for mobile applications, rovers would then be mobile reactors with minimal payload. Unless you have a better idea electric would be the way to go. Maybe redux flow battery.

**Comment by mokun** [comment]
> But that is no solution for mobile applications, rovers would then be mobile reactors with minimal payload. Unless you have a better idea electric would be the way to go. Maybe redux flow battery.

I haven't researched redux flow battery yet. Guess it's easy to make it space-rated. 

I suppose the whole supply/production chain, namely, the generation, handling, use of methanol, etc. has been reasonably investigated or hasn't passed all the milestones needed to be Mars-bound yet.

So it's still fair game when it comes to what type of power plant will be adopted for the 1st generation of a human rated Mars mission.

**Comment by mokun** [comment]
btw, NASA has just in recent years been pushing for the use of Radioisotope Thermoelectric Generator (RTGs) as mini power plant for robotic rovers.

It requires less plutonium to begin with, when comparing to a full nuclear reactor.

Still, we need to refine more plutonium on earth. 

Hopefully, we'll test, build and package them into future human occupied rovers prototypes and soon send them to Mars.

**Comment by Urwumpe** [comment]
RTGs are low power though. The rover needs a battery, to store energy for peak power (And you also can't turn it off). But RTGs can use more fuels than just Plutonium, which might be an option to consider. Plutonium is just the standard in space, because it lasts longer. Also you can't store RTGs for later use, when their fuel had been produced, the clock ticks.

**Comment by mokun** [comment]
> Also you can't store RTGs for later use, when their fuel had been produced, the clock ticks.

Yay. We can't simply start and stop RTGs right ?

> The rover needs a battery, to store energy for peak power (And you also can't turn it off)

I suppose with RTGs, at least we'll plug that into a battery to store power. 

Say a low power 10kW RTG is flight ready. But rovers already need more than 10kW to climb up hill, ride on rugged roads, or to accelerate.

btw, we've recently added regen braking in vehicles to get some power back.

We also modified VehicleSpec to allow provisions for multiple forms of power for providing propulsive energy.

Let's see where imagination takes us next.

**Comment by Urwumpe** [comment]
> Yay. We can't simply start and stop RTGs right ?

Exactly. The best way to make reliable technology. Give humans less options to mess with it.

**Comment by mokun** [comment]
btw, yesterday I saw a failure of a drone in reaching a destination settlement's supposed to be just 800 km away.

It flew 3000 km and went out of fuel and ended up "parking" somewhere.

We'll need to debug to find out how often a delivery mission is successful.

What do we do to a stranded drone ? Lost forever ?

**Comment by bevans2000** [comment]
Another case whether it is worth fixing if we are going to change Missions

**Comment by mokun** [comment]
I've found some bugs on DroneMission, causing TaskPhase to go null. Fixing now.

**Comment by mokun** [comment]
I'm getting a much better hang on the fuel economy/consumption of a drone now.

Here is a comparison of the energy density of 4 fuel sources in kWh per kg:

1. Jet fuel (avg. lower heating value of 43.147 MJ/kg, density of 0.809 kg/L):
Convert MJ/kg to kWh/kg: 43.147 MJ/kg × (1 kWh / 3.6 MJ) ≈ 12.01 kWh/kg

2. Methanol (not explicitly mentioned in the search results, but typical energy density is around 15.6 MJ/kg):
Convert MJ/kg to kWh/kg: 15.6 MJ/kg × (1 kWh / 3.6 MJ) ≈ 4.33 kWh/kg

3. Methane (not explicitly mentioned in the search results, but typical energy density is around 55.6 MJ/kg):
Convert MJ/kg to kWh/kg: 55.6 MJ/kg × (1 kWh / 3.6 MJ) ≈ 15.44 kWh/kg

4. Hydrogen (ΔG = 113 MJ/kg if water vapor is produced, and 118 MJ/kg if liquid water is produced):
Convert MJ/kg to kWh/kg: 113 MJ/kg × (1 kWh / 3.6 MJ) ≈ 31.39 kWh/kg (water vapor) or 32.78 kWh/kg (liquid water)

Glad that we won't have Jet fuel on Mars since there aren't any fossils.

Currently, our vehicle fleets uses methanol as methanol is logistically easier to handle than methane and hydrogen, at least on earth. Not sure if it's the same on Mars.

We do have a energy/power density problem. 

Delivery drones that rely on methanol can travel only short distances.

We'll need to switch to methane and hydrogen for longer hauling of cargoes.

To solve that, I'll add ***Cargo Drone*** that will use methane.

**Comment by bevans2000** [comment]
I think one solution is to create the concept of an Propulsion which all Vehicles have. Then we create different subclasses for the different propulsion approaches. Then it comes down to a configuration in the vehicles.xml as to what type a vehicle uses.
Plus we can create separate unit test to check each verient

**Comment by Urwumpe** [comment]
I would just make the abstract types of engines subclasses, since we could describe the finer types of engines could be described by attributes. This reduces the amount of engine subclasses and makes testing easier, since we could create any theoretical engine for testing the behaviour without caring for realism.

For example:

- [Heat engine](https://en.wikipedia.org/wiki/Heat_engine)
- Electric engine
- [Reaction engine](https://en.wikipedia.org/wiki/Reaction_engine)

Maybe we need to break down heat engine into more specific subtypes eventually, if we want more complex behaviour, but on the most abstract level, they behave all the same. 

The question is: Which outputs do we need for the vehicle engine? Speed and torque? Heat?

**Comment by mokun** [comment]
> Then we create different subclasses for the different propulsion approaches. 

Yay. We can use super class and subclass for various propulsion methods. 

Currently, the one big VehicleController class is not that good.

Also, we have Flyer class as the super class. Drone class is its subclass. They can in principles be used for adopting unique propulsion by composition.

**Comment by mokun** [comment]
> since we could create any theoretical engine for testing the behaviour without caring for realism.

Hmmm. The concept of implementing various types of engine sound like an interesting engineering job position for prospective applicants to become settlers on Mars :D 

> Which outputs do we need for the vehicle engine? Speed and torque? Heat?

So far, we accounted for actual fuel and energy consumption. We haven't considered the amount of heat it would produce yet.

Distance, time, speed, acceleration, force, power are quantities we have included in our models.

We've considered thrust but not torque exactly.

**Comment by mokun** [comment]
> [Heat engine](https://en.wikipedia.org/wiki/Heat_engine)

Combustion can one day become important for high power transportation. I heard people talked about using combustion engines on Mars

For now, we proudly use fuel cell stack and battery for low/mid consumption.

> [Reaction engine](https://en.wikipedia.org/wiki/Reaction_engine)

That we'll consider once we have re-invented rocket launches in mars-sim.

**Comment by mokun** [comment]
> fuel cells are not much different to batteries

but at least with fuel cell systems w need to store reactants onboard and all the associated logistic needs that go with it.

For battery systems, we need not worry about tanks and pumps, etc. We do need a robust charging infrastructure.

I see this though-provoking pic that compares the system efficiencies across the battery, hydrogen, diesel and petrol approach : 
![electric vs hydrogen overall efficiency](https://github.com/user-attachments/assets/b3c6b591-9209-4eca-bbf0-7ba0d3e9826c)

**Comment by bevans2000** [comment]
> > For now, we proudly use fuel cell stack and battery for low/mid consumption.
> 
> Well, fuel cells are not much different to batteries, especially batteries with rechargable electrolyte. Maybe we can cover them with just one class.

If we want to take it one step further we should have a ```Battery``` class that can be shared with ```PowerStorage``` as well as in the Vehicle

**Comment by mokun** [comment]
> If we want to take it one step further we should have a Battery class that can be shared with PowerStorage as well as in the Vehicle

yay it's coming but will take some time to synchronize the battery codes between the vehicle, the building and the robot.
